src/test/scala/uk/gov/nationalarchives/testUtils/ExternalServicesTestUtils.scala

```diff
@@ -0,0 +1,126 @@
+    when(asyncClientMock.listObjectsV2(any[ListObjectsV2Request]))
+      .thenReturn(CompletableFuture.completedFuture(listObjectsV2Response))
+
+    IO(listObjectsV2Publisher.commonPrefixes().map(_.prefix()))
+  }
+
+  case class MockLambda(sdkPublisher: IO[SdkPublisher[String]], s3UploadResult: IO[CompletedUpload]) extends Lambda() {
+    private val mockS3Client: DAS3Client[IO] = mock[DAS3Client[IO]]
+    override val dAS3Client: DAS3Client[IO] = {
+      when(
+        mockS3Client.listCommonPrefixes(
+          any[String],
+          any[String]
+        )
+      ).thenReturn(sdkPublisher)
+      when(
+        mockS3Client.upload(
+          any[String],
+          any[String],
+          any[Long],
+          any[Publisher[ByteBuffer]]
+        )
+      ).thenReturn(s3UploadResult)
+      mockS3Client
+    }
+
+    def verifyInvocationsAndArgumentsPassed(
+        foldersInManifest: List[String] = Nil,
+        numberOfUploads: Int
+    ): Unit = {
+      val stagingCacheBucketCaptor: ArgumentCaptor[String] = ArgumentCaptor.forClass(classOf[String])
+      val keysPrefixedWithCaptor: ArgumentCaptor[String] = ArgumentCaptor.forClass(classOf[String])
+
+      val uploadBucketCaptor: ArgumentCaptor[String] = ArgumentCaptor.forClass(classOf[String])
+      val keyToUploadCaptor: ArgumentCaptor[String] = ArgumentCaptor.forClass(classOf[String])
+      val opexFileContentLengthCaptor: ArgumentCaptor[Long] = ArgumentCaptor.forClass(classOf[Long])
+      val publisherCaptor: ArgumentCaptor[StreamUnicastPublisher[IO, ByteBuffer]] =
+        ArgumentCaptor.forClass(classOf[StreamUnicastPublisher[IO, ByteBuffer]])
+
+      verify(mockS3Client, times(1)).listCommonPrefixes(
+        stagingCacheBucketCaptor.capture(),
+        keysPrefixedWithCaptor.capture()
+      )
+
+      stagingCacheBucketCaptor.getValue should be("stagingCacheBucketName")
+      keysPrefixedWithCaptor.getValue should be("opex/9e32383f-52a7-4591-83dc-e3e598a6f1a7/")
+
+      verify(mockS3Client, times(numberOfUploads)).upload(
+        uploadBucketCaptor.capture(),
+        keyToUploadCaptor.capture(),
+        opexFileContentLengthCaptor.capture(),
+        publisherCaptor.capture()
+      )
+
+      if (numberOfUploads > 0) {
+        val keysToUpload: String = keyToUploadCaptor.getValue
+        val publisher: StreamUnicastPublisher[IO, ByteBuffer] = publisherCaptor.getValue
+
+        uploadBucketCaptor.getValue should be("stagingCacheBucketName")
+        keysToUpload should equal("opex/9e32383f-52a7-4591-83dc-e3e598a6f1a7/9e32383f-52a7-4591-83dc-e3e598a6f1a7.opex")
+        opexFileContentLengthCaptor.getValue should be(346)
+
+        val opexInByteBuffer: Array[ByteBuffer] = publisher.stream.compile.toList.unsafeRunSync().toArray
+        val opexAsArrayOfBytes = opexInByteBuffer.flatMap(_.array())
+        val opexAsString = opexAsArrayOfBytes.map(_.toChar).mkString
+
+        opexAsString should equal(s"""<opex:OPEXMetadata{SPACE_THAT_INTELLIJ_REMOVES}
```

Comment:
If you have this as an XML `Elem` and then use PretyPrinter, you can get rid of all the replace calls. 
```scala
        val prettyPrinter = new PrettyPrinter(80, 2)
        val expectedXml =
        <opex:OPEXMetadata xmlns:opex="http://www.openpreservationexchange.org/opex/v1.2">
          <opex:Transfer>
            <opex:Manifest>
          ...
         opexAsString should equal(prettyPrinter.format(expectedXml))
```